lib/parser/Parser.cpp

```diff
@@ -0,0 +1,1609 @@
+        if (success == 1) {
+          node.add(child_node);
+          child_node.reset();
+          if (tokens[current_token_index].token_type == TOKEN_SEMICOLON) {
+            child_node.set_type(leaf);
+            child_node.set_token(tokens[current_token_index]);
+            node.add(child_node);
+            child_node.reset();
+            advance(tokens);
+            return 1;
+          } else {
+            child_node.reset();
+            node.reset();
+            current_token_index = current_token_index_placeholder;
+          }
+        } else {
+          child_node.reset();
+          node.reset();
+          current_token_index = current_token_index_placeholder;
+        }
+      } else {
+        child_node.reset();
+        node.reset();
+        current_token_index = current_token_index_placeholder;
+      }
+    } else {
+      child_node.reset();
+      node.reset();
+      current_token_index = current_token_index_placeholder;
+    }
+  } else {
+    child_node.reset();
+    node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  return 0;
+}
+
+int Parser::parse_entryStatement(std::vector<Token> tokens, Node &node) {
+  int success = 0;
+  int current_token_index_placeholder = current_token_index;
+
+  node.set_type(entryStatement);
+  Node child_node;
+
+  success = parse_registerDeclaration(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  success = parse_location(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  success = parse_label(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  success = parse_pragma(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  success = parse_callprototype(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+
+  success = parse_calltargets(tokens, child_node);
+  if (success == 1) {
+    node.add(child_node);
+    return 1;
+  } else {
+    child_node.reset();
+    current_token_index = current_token_index_placeholder;
+  }
+  return 0;
+}
+
+int Parser::parse_registerDeclaration(std::vector<Token> tokens, Node &node) {
+  int success = 0;
+  int current_token_index_placeholder = current_token_index;
+  node.set_type(registerDeclaration);
+  Node child_node;
+
+  if (tokens[current_token_index].token_type == TOKRN_REG) {
```

Comment:
A lot of these have a pattern of successively applying parse functions. Maybe there is a way to refactor them.

Reply:
I can address all your other comments, but im sure what you mean by this one